modernrpc/handlers/jsonhandler.py

```diff
@@ -87,6 +87,8 @@ def process_request(self):
                     # Response objects for notifications."
                     if request_id:
                         batch_result.results.append(self.json_success_response(result, override_id=request_id))
+                except AuthenticationFailed as e:
+                    raise e
 
```

Comment:
First, AuthenticationFailed inherits from RPCInternalError which itself inherits from RPCException.
As a result, such kind of exception is already catched by the next except block.

Second, as you may have noticed, the batch result must return a valid JSON payload with information about encountered error. You cannot just re-raise the exception, breaking the result of the whole batch returned to client.

Reply:
In that case, the behavior is different for single calls / batch calls (in terms of authentication):
- Single call -> 403 Forbidden
- Batch call -> 200 Ok

Reply:
Indeed, both XML-RPC and JSON-RPC are pretty flexible on HTTP status code returned.

XML-RPC states that (see http://xmlrpc.com/spec.md section "Response format"):
> Unless there's a lower-level error, always return 200 OK.

JSON-RPC, as it was designed to be compatible with any protocol (not only HTTP) (see https://www.jsonrpc.org/specification#overview) doesn't imply a specific status code or another (see https://www.jsonrpc.org/specification#response_object). But the behavior of batch request specifically mention that the server must return an array with all results, including error payloads when needed (see https://www.jsonrpc.org/specification#batch)

I understand that behavior is not ideal but it is most a limitation of both standards than an implementation issue.

Reply:
> But the behavior of batch request specifically mention that the server must return an array with all results, including error payloads when needed

This is also valid for single calls (https://www.jsonrpc.org/specification#response_object):

> When a rpc call is made, the Server MUST reply with a Response, except for in the case of Notifications. The Response is expressed as a single JSON Object, with the following members:

With these two guidelines (single/batch) in mind, a 200 status code should always be returned along with a valid JSON response. And avoid this response customization (in `views.py`):

```py
except AuthenticationFailed as e:
    # Customize HttpResponse instance used when AuthenticationFailed was raised
    logger.warning(e)
    return handler.result_error(e, HttpResponseForbidden)
```

What do you think?

Reply:
No, since JSON-RPC standard was written to work on top of any protocol (not only HTTP), the status code is never mentioned and is irrelevant. I dan't understand how to deduce the 403 status must be changed from the sentences you quote. Actually, your original PR didn't solve the problem you described.

But I agree that the library should allow user to customize that behavior and return a 200 instead of a 403 in case of AuthenticationError. I could add a new setting to allow such a customization?

Would you agree with that ?

Reply:
That would be great.

Thanks a lot!